openfl/federated/data/sources/s3_data_source.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2020-2025 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import hashlib
+import os
+from typing import Generator
+from urllib.parse import urlparse
+
+import boto3
+
+from openfl.federated.data.sources.data_source import DataSource, DataSourceType
+
+
+class S3DataSource(DataSource):
+    """Class for S3 data"""
+
+    def __init__(
+        self,
+        uri: str,
+        endpoint=None,
+        aws_access_key_id=None,
+        aws_secret_access_key=None,
+        hash_func=None,
+    ):
+        super().__init__(DataSourceType.S3)
+        self.uri = uri
+        self.endpoint = endpoint
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.hash_func = hash_func
+
+    def enumerate_files(self) -> Generator[str, None, None]:
+        """Enumerate all files in the data source"""
+        parsed = urlparse(self.uri)
+        bucket_name = parsed.netloc
+        prefix = parsed.path.lstrip("/")  # Remove leading slash
+
+        aws_access_key_id = (
+            os.environ.get(self.aws_access_key_id) if self.aws_access_key_id else None
+        )
+        aws_secret_access_key = (
+            os.environ.get(self.aws_secret_access_key) if self.aws_secret_access_key else None
+        )
+
+        s3_client = boto3.client(
```

Comment:
can be moved to __init__ once and for all

Reply:
nice, done